transport_plugins/bled112/iotile_transport_bled112/async_packet.py

```diff
@@ -51,37 +60,39 @@ def read_packet(self, timeout=3.0):
             raise InternalTimeoutError("Timeout waiting for packet in AsyncPacketBuffer")
 
 
-def ReaderThread(filelike, read_queue, header_length, length_function, stop):
+def reader_thread(filelike, read_queue, header_length, length_function, stop, dedupe=False, dedupe_timeout=5):
     logger = logging.getLogger(__name__)
+    broadcast_v2_dedupers = None
+    read_buffer = bytearray()
+
+    if dedupe:
+        broadcast_v2_dedupers = BroadcastV2DeduperCollection(dedupe_timeout)
 
     while not stop.is_set():
         try:
-            header = bytearray()
-            while len(header) < header_length:
-                chunk = bytearray(filelike.read(header_length - len(header)))
-                header += chunk
+            # The bled112 will read EOF when there is no more data, so it's safe to read a large amount
+            read_buffer += bytearray(filelike.read(1024))
```

Comment:
This read pattern is not a safe way to optimize this.  It trades speed in some use cases for significant slowdown and uncontrolled performance in many others.  

I suspect that you should see that if you do a test where you send RPCs or a script to a POD, you'll find the throughput is now significantly slower with this code (possibly as much as 3-4x slower) and highly variable based on how many PODs are nearby sending advertisement traffic.

This issue is that `filelike.read(1024)` blocks until either of the following conditions are met:
- there are at least 1024 bytes available OR
- the configured read timeout (set at 100ms I believe) expires and whatever bytes it has are returned

In scenarios when you are writing a command and reading the response before writing the next command, you will never have 1024 bytes filling up the buffer quickly so every command you write will now take a minimum of 100ms because you have to wait for the read to timeout before the several byte response to the command is returned to you.

In an RPC we need 2 writes (write payload, write command) and normally that takes < 1ms followed by 15-30ms for those writes to get across BLE to the POD and the response to come back.  With this change that would become 200 ms worst case for the writes since it needs to two timeouts.

The safe pattern for this kind of thing:
```python
   read_buffer += filelike.read(1)  # Is there any data at all available without assuming how much?
 
   avail_length = filelike.in_waiting  # NOTE That this is a pyserial api, not generic filelike so you may need to modify the tests

  if avail_length > 0:
      read_buffer += filelike.read(avail_length)  # This can never timeout because you just guaranteed there's enough data
```

Reply:
Sounds good. I was a bit surprised when I didn't notice any performance issues, and I most likely didn't perform the right test to see them. `in_waiting` looks great, I'll implement that.

Reply:
Confirmed, pulling a large amount of text from a device, using an RPC sequence that requires one RPC request and response per 20 bytes of text, took about 20% longer with the current code. (though not 3-4x slower as predicted). Tests where both the sender and reciver used the code from this PR took 30% longer. Will test with the changes @timburke suggested.

Reply:
The adjustment looks good, I'm seeing similar times to read 5500 characters from an AP on all versions again. 

I'm working on upgrading my performance tests so I'll have to wait to see how this affects advertisement performance.